### ISSUE TITLE
Add URL Parameter Tracking

### ISSUE THREAD

**Issue by aciuksza** [fpcorso/quiz_master_next#373]
Hello, Frank!

I've seen a couple of other systems with this feature and would really benefit from seeing it in QMN.

In short, I'd love to be able to pass a URL parameter for tracking purposes (i.e. http://quiz.com/quizurl?s=sourcecode) that can populate into the database upon submission of a quiz.

Use case -- we use this system for assessments that audience members can take during one of our executive keynotes. We would like to see what conversion rates look like per event, what the data looks like generally from the event, and how things compare event-to-event.

At one point, I was able to do this by making some changes to the code base and inputting the parameter into the "Business" section of the table, but that was purely for testing and it didn't have any way of stripping out malicious code.

Would this be something that could be done in the core or as an add-on?

-Albert

**Comment by fpcorso** [comment]
Hey @aciuksza,

Thank you for choosing to use this plugin! This is a really cool idea. Would this need to be something that could handle unlimited parameters or would one single parameter be enough? You mentioned a few other systems with these feature, could you mention one or two so that I could possibly take a look?

On the backend, would you want this in the Results table near the Email and Business columns or would this be more in the Results Details upon viewing the results?

**Comment by aciuksza** [comment]
Hey Frank:

For sure -- really looking forward to the implementation!

Here's an example: http://ipanelthemes.com/kb/fsqm/fsqm-other-features/tracking-submission-from-url/

To answer your questions...
1) For my use case, one would be absolutely sufficient. Not sure about what others might think could be useful.
2) On the backend, the results table (near email/business) would be great, but not a deal-breaker.

Thanks again,
Albert

**Comment by aciuksza** [comment]
Hey Frank -- seeing that this keeps getting bumped.

One of the reasons we're using SurveyGizmo right now (to the tune of about $900/year) is because I can pass data using multiple URL parameters. Would love to move to QMN, but can't until we have this feature.

They've structured it in a way that is simple to implement and could be done securely by filtering out any code injection. 

https://help.surveygizmo.com/help/url-merge-codes

In short, it's a hidden text box that defaults to a code structure that triggers a look at the URL to get the parameter. It's simple on implementation from the user perspective and I think it would probably be simple to implement on the code side.

Thoughts?

-Albert

**Comment by fpcorso** [comment]
Ah! Just found another issue that needs to be closed before this one is possible. Need to implement #622 to get this URL parameter created. Increasing that issue's priority to the active milestone (5.2.0) to get it out quicker.

**Comment by fpcorso** [comment]
Current initial mockup of the feature admin area:

![url-parameter-feature-mockup](https://user-images.githubusercontent.com/9730040/36355641-6a0567c6-14b4-11e8-86e1-d8ca65dbf15c.png)

By having the "parameter" and "label" as two separate values, it allows the user to set up a parameter such as "a=23" but the results show "Age: 23".

I wonder if any user would ever need one of the parameters to be required in order to access the quiz/survey... The next big item to ponder is if we should have values equal different labels. We could have the value be one of three and then attach labels to those value. For example, say we had referrrer:

Parameter "a", Label "referrer". We could have value 1 and 2 tied to labels "My site" and "My Other Site". So, the URL would say a=1 but the internal results could show "Referrer: My Site". This might be too big for the first version but would be interesting to consider for future updates if it doesn't make it into the first version.

**Comment by aciuksza** [comment]
Hey Frank:


Appreciate your thoughts here.

I really like the idea of being able to create a label for the referrer --
keeps things much, much cleaner.

Keep me posted -- this would have a huge impact on our ability to use the
tool.

Best,
Albert


---
*Albert B. Ciuksza Jr.*
412.596.0848
www.ciuksza.com

twitter.com/albertciuksza
facebook.com/albertciuksza
linkedin.com/in/albertciuksza

**Comment by fpcorso** [comment]
I've created a simple prototype of the admin based on the initial mockup:

![url-parameter-prototype](https://user-images.githubusercontent.com/9730040/36703691-dae8bc96-1b2a-11e8-9f36-dc4bd6ebd06a.png)

Clicking "create new" or the edit icon would open a popup for editing the parameter much like how the new question editor in 5.2 is going to work. Still working on how this will work on the results side but I have a few ideas to sketch up. Ideally, this would be added as a filterable field in the Reporting & Analysis addon, so users could aggregate results based on certain parameters.

**Comment by fpcorso** [comment]
Based on the usage and how this will work, I am leaning more towards addon instead of building this into core but haven't decided one way or the other yet. I try to add to the core if more than 30% of users would use a feature. 

However, I don't think this feature will have that amount of adoption based on our data of how the plugin is used.

**Comment by aciuksza** [comment]
Hey Frank:

Works either way for me. Makes sense if you don't feel that it's a
commonly-used core function.

Best,
Albert

---
*Albert B. Ciuksza Jr.*
412.596.0848
www.ciuksza.com

twitter.com/albertciuksza
facebook.com/albertciuksza
linkedin.com/in/albertciuksza

**Comment by fpcorso** [comment]
This has been converted into a future addon. I am closing this ticket here but I have emailed you, @aciuksza for further conversation.
